### PR TITLE
Add External Engine API methods

### PR DESCRIPTION
**Pull request by joshka**
Builds on #12 - so if reviewing just read the last 2 commits:
- feat: add external engine API methods https://github.com/ion232/lichess-api/pull/14/commits/5f7bef4a02d3d21e55358e4d9f53d0cc0e230ce8 
- feat: add engine analyse https://github.com/ion232/lichess-api/pull/14/commits/bc579558c61b55ebc2f3cb2eda5d45434ed26b24

WORKING:
- create
- update
- list
- get
- delete

TODO:
- /engine/{id}/analyse gives a 404 currently (https://lichess.org/api#tag/External-engine/operation/apiExternalEngineAnalyse)
- https://lichess.org/api#tag/External-engine/operation/apiExternalEngineAcquire
- https://lichess.org/api#tag/External-engine/operation/apiExternalEngineSubmit (requires streaming POST - so this might be a bit more effort than the other methods)

### CODE REVIEW

**Comment by joshka** [comment]
Rebased on main and cleaned up the commits - I haven't looked at this in a while, so I've forgotten where it was up to. Just pushing what I had locally

**Comment by ion232** [comment]
I've just pushed some changes that add support for other domains in a way that's different to yours. The domain is now part of the request itself, but defaults to lichess.org, so it isn't really noticeable unless you need to change it. You'll have to update your code to use that for the `engine.lichess.ovh` endpoints

**Comment by joshka** [comment]
Neat - thanks for that :)

File: Cargo.toml [hunk 1]
```diff
@@ -33,3 +33,7 @@ clap = { version = "4.4.18", features = ["derive"] }
 color-eyre = "0.6.2"
 tokio = { version = "1.35.1", features = ["macros", "rt-multi-thread"] }
 tracing-subscriber = { version = "0.3.18", features = ["env-filter"] }
+rand = "0.8.4"
+
+# [features]
+# awc = ["dep:actix-web", "dep:awc", "dep:webpki-roots", "dep:mime", "dep:rustls"]
```
Comment (ion232) [inline Cargo.toml:39]:
```suggestion
```

File: examples/lichess.rs [hunk 2]
```diff
@@ -139,3 +146,227 @@ impl PuzzleCommand {
...
+//       "infinite": false,
+//       "multiPv": 1,
+//       "variant": "chess",
+//       "initialFen": "rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1",
+//       "moves": [
+//         "e2e4",
+//         "g8f6"
+//       ]
+//   }
+
```
Comment (ion232) [inline examples/lichess.rs:233]:
```suggestion
```

File: src/api/external_engine.rs [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+        request: list::GetRequest,
+    ) -> Result<Vec<ExternalEngine>> {
+        self.get_single_model(request).await
+    }
+
+    pub async fn create_external_engine(
+        &self,
+        request: create::PostRequest,
+    ) -> Result<ExternalEngine> {
+        debug!("Creating external engine: {:?}", request);
```
Comment (ion232) [inline src/api/external_engine.rs:20]:
```suggestion
```

File: src/api/external_engine.rs [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+
+    /// This method currently returns a 503 error (Service Unavailable) from the Lichess API
+    pub async fn analyse_with_external_engine(
+        &self,
+        request: analyse::PostRequest,
+    ) -> Result<impl StreamExt<Item = Result<analyse::AnalysisResponse>>> {
+        self.get_streamed_models(request).await
+    }
+
+    pub async fn aquire_analysis_request(
```
Comment (ion232) [inline src/api/external_engine.rs:47]:
```suggestion
    pub async fn acquire_analysis_request(
```

File: examples/lichess.rs [hunk 5]
```diff
@@ -139,3 +146,227 @@ impl PuzzleCommand {
...
+                provider_secret,
+                wait,
+            } => {
+                let acquire_analysis = acquire_analysis::AcquireAnalysis { provider_secret };
+                let request = acquire_analysis::PostRequest::new(acquire_analysis);
+                let mut analysis = lichess.aquire_analysis_request(request.clone()).await?;
+                while wait && analysis.is_none() {
+                    tracing::debug!("No analysis request available");
+                    analysis = lichess.aquire_analysis_request(request.clone()).await?;
+                }
```
Comment (ion232) [inline examples/lichess.rs:355]:
```suggestion
                let mut analysis = lichess.acquire_analysis_request(request.clone()).await?;
                while wait && analysis.is_none() {
                    tracing::debug!("No analysis request available");
                    analysis = lichess.acquire_analysis_request(request.clone()).await?;
                }
```

**Review by ion232 — CHANGES_REQUESTED** [review]
Hi @joshka

Thanks for your work on the external engine. I'd like to get it into the `main` branch so it's not diverging while other changes are going in. I'll look into finishing it off once its in there. I've suggested a few bits of cleanup before it's ready.

**Review by ion232 — APPROVED** [review]

**Pull request merged**